WPMedia/fusion-news-theme-blocks
# TMEDIA-289 - Review Results List for React - Optimization

## Description
Rewrite block to use functional components.  Rewrote tests using `@testing-library/react`

## Jira Ticket
- [TMEDIA-289](https://arcpublishing.atlassian.net/browse/TMEDIA-289)

## Acceptance Criteria
Optimize React Render methods

## Test Steps
- Add test steps a reviewer must complete to test this PR

0. Run `npm i` for additional packages around testing
1. Checkout this branch `git checkout TMEDIA-289-result-list-refactor`
2. Run fusion repo with linked blocks `npx fusion start -f -l @wpmedia/results-list-block`
3. Create a new page and add the results-list-block to it
4. Configure the results-list-block as follows:
![image](https://user-images.githubusercontent.com/2287238/124684307-cec1bd80-de9c-11eb-8fa3-0ca4652ab133.png)
5. You should see it render the `/sara-carothers` articles 5 at a time.
6. Configure the results-list-block as follows:
![image](https://user-images.githubusercontent.com/2287238/124684661-a6868e80-de9d-11eb-947f-24ca93ceb5da.png)
7. You should see the homepage articles starting from the third article. (verify this by setting from to 0 and see the third)
8. Configure the following values and validate no button (I only have two stories for this, if you have more, increase size):
![image](https://user-images.githubusercontent.com/2287238/124686095-7bea0500-dea0-11eb-9135-716cfe57c3fa.png)
9. Page should preload 5 additional items, but not render them.  You can check this by viewing the network tab in debug and counting the number of items returned from the initial request.  Any press of the `See More` button will only pull the number of items equal to your size.
![image](https://user-images.githubusercontent.com/2287238/124686378-07639600-dea1-11eb-9020-d9fbbbf2fab2.png)
![image](https://user-images.githubusercontent.com/2287238/124686413-15b1b200-dea1-11eb-95a7-fa6691ff0716.png)
![image](https://user-images.githubusercontent.com/2287238/124686433-2104dd80-dea1-11eb-8ed8-1030b073a71b.png)

## Effect Of Changes
### Before
Pages would load content as requested, causing a slight delay to the user
Render passes (first and after hitting see more per @JackHowa's recommendation):
![image](https://user-images.githubusercontent.com/2287238/125114793-6d763600-e0b8-11eb-9baa-3bca6428fce9.png)
![image](https://user-images.githubusercontent.com/2287238/125114877-8979d780-e0b8-11eb-82e5-c8d9bc1d0530.png)
Lighthouse Result:
![image](https://user-images.githubusercontent.com/2287238/125114714-520b2b00-e0b8-11eb-838b-94eaff4e70fb.png)


### After
Pages should show additional content immediately and load more in the background
Render passes:
![image](https://user-images.githubusercontent.com/2287238/125113323-85e55100-e0b6-11eb-9f72-bc62712a3b9f.png)
![image](https://user-images.githubusercontent.com/2287238/125113400-9e556b80-e0b6-11eb-88f8-94f74331afa5.png)
Lighthouse Result:
![image](https://user-images.githubusercontent.com/2287238/125113236-60f0de00-e0b6-11eb-85d8-709df8ce9ef3.png)


## Dependencies or Side Effects
N/A

## Author Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
please explain why (so that we can fix it whenever it gets refactored).

## Reviewer Checklist
_The reviewer of the PR should copy-paste this template into the review comments on review._
